File: appl/02/circle.rs
use std::fmt;

struct Point(f32, f32);  // tuple struct

struct Vector(f32, f32);

struct Circle {
    center: Point,
    radius: f32,
}

impl Circle {
    fn new(center: Point, radius: f32) -> Self {
        Self {
            center,
            radius,
        }
    }

    fn scale(&mut self, factor: f32) {
        self.radius *= factor;
    }

    fn translate(&mut self, v: Vector) {
        self.center.0 += v.0;
        self.center.1 += v.1;
    }

    fn area(&self) -> f32 {
        std::f32::consts::PI * self.radius * self.radius
    }
}

impl fmt::Display for Circle {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "({}, {})", self.center, self.radius)
    }
}

impl fmt::Display for Point {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "({}, {})", self.0, self.1)
    }
}

fn main() {
    let mut c = Circle::new(Point(0.0, 0.0), 1.0);
    println!("{}", c.area());
    println!("{}", Circle::area(&c));

    c.scale(2.0);
    println!("{}", c.area());

    Circle::scale(&mut c, 0.5);
    println!("{}", Circle::area(&c));

    println!("{c}");
}

File: cp/07/circle.go
package main

import "fmt"

type Circle struct {
  x, y, r int
}

func (c Circle) Draw() {  // value receiver
  fmt.Printf("(%d, %d), %d\n", c.x, c.y, c.r)
}

func (c *Circle) Scale(factor int) {  // pointer receiver
  c.r *= factor
}

func main() {
  c1 := Circle{0, 0, 1}
  c2 := Circle{
    x: 1,
    y: 2,
    r: 3,
  }

  c1.Draw()
  c2.Scale(5)
  c2.Draw()
}
